### ISSUE TITLE
Rule to mandate/forbid trailing commas

### ISSUE THREAD

**Issue by 3flex** [pinterest/ktlint#709]
## Expected Behavior
ktlint offers a rule that enforces a particular code style regarding trailing commas, which are now permitted by the Kotlin compiler as of 1.3.70, see https://youtrack.jetbrains.com/issue/KT-34743 for more details.

This may not be a good fit for ktlint as it's too opinionated and no conventions exist yet, but if it's suitable it's a better fit here than in detekt where [it was originally proposed](https://github.com/detekt/detekt/issues/2398).

## Observed Behavior
No such rule exists.

## Steps to Reproduce
N/A, new rule required.

## Your Environment

* Version of ktlint used:
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task):
* Version of Gradle used (if applicable):
* Operating System and version:
* Link to your project (if it's a public repository):

**Comment by alexvanyo** [comment]
Ideally this rule could be as flexible as something like [`ESLint`'s comma dangling rule](https://eslint.org/docs/rules/comma-dangle), which would allow specifying when trailing commas must be used (never, always, multiline-only, `when` entry, etc.)

**Comment by Tapchicoma** [comment]
Probably related to #701 to control `enable/disable` state for cases, though initial rule implementation could just enforce trailing commas and be in experimental set.

**Comment by s1monw1** [comment]
Any update?

**Comment by ivorsliide** [comment]
Is breaking so many things this  ,
https://stackoverflow.com/questions/65055798/kotlin-linting-rulesetprovider-trailing-comma

**Comment by 3flex** [comment]
Is this on ktlint's roadmap? If not, we could implement in detekt instead. I don't think it makes sense to do that though if ktlint will implement the rule so we can avoid duplication of effort, but I think there's demand for it (I'm sure this is my most 👍 issue ever!)

**Comment by ivorsliide** [comment]
Hopefully is or it will be 🤞, for now I'm trying to implement my own rule from the link above. I can confirm there is a high demand. Will be nice to autocorrect with ktlint...

**Comment by romtsn** [comment]
@3flex you are more than welcome to contribute this to ktlint directly and utilize it in detekt afterwards 👀   it is certainly on the roadmap, just the maintainers were sidetracked by the day jobs and we had other priorities in ktlint, but I see this one indeed has a high demand so we'll re-prioritize

**Comment by CDehning** [comment]
Is this is getting released soon?

**Comment by ivorsmorenburg** [comment]
> Is this is getting released soon?

Maybe try using your own ruleset
- https://stackoverflow.com/questions/65055798/kotlin-linting-rulesetprovider-trailing-comma

**Comment by Badneighbour91** [comment]
Oui oumême iii

**Comment by snijsure** [comment]
> > Is this is getting released soon?
> 
> Maybe try using your own ruleset
> https://stackoverflow.com/a/65449838/3438335

Is there a guideline on integrating this rule into version of local ktlint?

**Comment by christian-draeger** [comment]
any news?

**Comment by TobiasMende** [comment]
We are currently using our own rule for mandating (and automatically fixing) missing trailing commas. Probably it is not ready yet to be added to the ktlint project, but I am happy to have a discussion about it:

```
class MandateTrailingComma : Rule("mandate-trailing-comma") {

    override fun visit(
        node: ASTNode,
        autoCorrect: Boolean,
        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
    ) {
        if (node.isArgumentOrParameterList() && node.isMultiLine()) {
            val lastArgument = node.lastArgumentOrParameter()
            if (lastArgument != null && lastArgument.treeNext.elementType != COMMA) {
                emit(
                    lastArgument.startOffset + lastArgument.textLength,
                    "Trailing comma is missing",
                    true,
                )
                if (autoCorrect) {
                    node.addChild(LeafPsiElement(COMMA, ","), lastArgument.treeNext)
                }
            }
        }
    }

    private fun ASTNode.isArgumentOrParameterList(): Boolean =
        elementType == VALUE_ARGUMENT_LIST || elementType == VALUE_PARAMETER_LIST

    private fun ASTNode.isMultiLine(): Boolean = children().any { it is PsiWhiteSpace && it.textContains('\n') }
    private fun ASTNode.lastArgumentOrParameter() =
        children().findLast { it.elementType == VALUE_ARGUMENT || it.elementType == VALUE_PARAMETER }
}
```

**Comment by romtsn** [comment]
Hey everyone following this thread - apologies for taking this so long, but it's finally there 🎉 Will be available with the next release, please try out the snapshot build in the meantime. Note, that it's available as a `--experimental` rule for now.

We tried to make it work the same way as IntelliJ does, meaning, depending on the editorconfig option
```Editorconfig
ij_kotlin_allow_trailing_comma = true|false
ij_kotlin_allow_trailing_comma_on_call_site = true|false
``` 
the rule will report either missing or redundant commas and fix them. The IntelliJ/AS itself should pick this up, so there should be no divergences between hitting the `Option+Cmd+L` and running ktlint (otherwise it's a bug worth reporting).

**Comment by sindrenm** [comment]
Fantastic stuff! Just to be clear, “allowing” trailing commas here means _requiring_ trailing commas, right? Will

```kotlin
fun foo(
  param1: Bar,
  param2: Bar
)
```

be converted into

```kotlin
fun foo(
  param1: Bar,
  param2: Bar, // <-- added trailing comma
)
```

when either running `ktlint -F` or the auto-formatter in IJ/AS?

**Comment by romtsn** [comment]
Yes, this means **requiring**, that's correct (and what IJ does as well, when ticking the "Allow trailing comma" checkbox in settings). I will update README with a better description of the behavior 👍

**Comment by snijsure** [comment]
I might be doing something wrong, and I don't see ktlint -F is putting missing trailing commas,

Here is my setup:

I have the following in .editorconfig

```
ij_kotlin_allow_trailing_comma = true
ij_kotlin_allow_trailing_comma_on_call_site = true
```

And I have code like this

```
    data class State(
        val tutorialPages: List<TutorialPage> = emptyList(),
        val currentPageNumber: Int = 0
    ) 

```

When I run command

```
ktlint -F --editorconfig=.editorconfig --experimental fileName
```

I see the following error

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jetbrains.kotlin.com.intellij.util.ReflectionUtil (file:/usr/local/Cellar/ktlint/0.42.1/libexec/ktlint) to field java.lang.Throwable.backtrace
WARNING: Please consider reporting this to the maintainers of org.jetbrains.kotlin.com.intellij.util.ReflectionUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

I don't see that ktlint -F has put the trailing comma.

```
  data class State(
        val tutorialPages: List<TutorialPage> = emptyList(),
        val currentPageNumber: Int = 0**,**
    ) 
```

**Comment by romtsn** [comment]
@snijsure which version are you using? `ktlint --version`

**Comment by snijsure** [comment]
@romtsn I am using version 0.42.1
```
ktlint --version
0.42.1
```

**Comment by romtsn** [comment]
yeah, as I mentioned it's only available in the [snapshot](https://github.com/pinterest/ktlint#access-to-the-latest-master-snapshot) version for now.

**Comment by zsmb13** [comment]
A new release containing this was just created https://github.com/pinterest/ktlint/releases/tag/0.43.0

**Comment by mohsenoid** [comment]
I am using 0.43.0 CLI
![image](https://user-images.githubusercontent.com/5135535/144238379-1b93cfff-66c5-4408-ada6-493954eaa0fa.png)
and enabled the following in my project `.editorconfig`
```
ij_kotlin_allow_trailing_comma = true
ij_kotlin_allow_trailing_comma_on_call_site = true
```
but I don't get any format or lint error from ktlint when I execute this:
```ktlint -F "app/src/**/*.kt"```

On the other hand, the Android studio code styler understands it and does the formatting for me, but I also want it to be a part of the CI check!

**Comment by romtsn** [comment]
@mohsenoid 
> Note, that it's available as a --experimental rule for now.

**Comment by snijsure** [comment]
Is there a way to just enable this rule, out of the experimental ruleset?

 If I turn on experimental rules it turns up lots of other warnings that we don't care at this moment

@romtsn

**Comment by mohsenoid** [comment]
@romtsn Thanks for the hint, I executed this and it works fine:
```ktlint -F --experimental "app/src/**/*.kt"```

**Comment by romtsn** [comment]
@snijsure nope, the best you can do is to enable the experimental ruleset and disable the rules you don't care about via `disabled_rules` option

**Comment by mohsenoid** [comment]
Since this is an experimental rule, should we report issues here or create a new ticket case by case?

For instance in this case
https://github.com/mohsenoid/Rick-and-Morty/blob/c9fb8dc093c7986055e8892b8e126f60bf0247cc/app/src/main/kotlin/com/mohsenoid/rickandmorty/data/db/DbRickAndMorty.kt#L51

Android Studio works correctly and adds a comma, but Ktlint doesn't and even removes the comma added by Android Studio!

**Comment by romtsn** [comment]
@mohsenoid please file a new issue

**Comment by knyttl** [comment]
how can I enable this rule via .editorconfig? I used:

```
ij_kotlin_allow_trailing_comma = true
ij_kotlin_allow_trailing_comma_on_call_site = true
```

but seems to have no effect

**Comment by romtsn** [comment]
https://github.com/pinterest/ktlint/issues/709#issuecomment-983649250

**Comment by knyttl** [comment]
> #709 (comment)

It does not answer how I can enable it via .editorconfig.

**Comment by romtsn** [comment]
well, you question was about the trailing comma rule, not about the experimental ruelset. you can't enable experimental via editorconfig

**Comment by Egorand** [comment]
Can this feature be used with [spotless](https://github.com/diffplug/spotless/tree/main/plugin-gradle#ktlint)?

**Comment by sindrenm** [comment]
> Can this feature be used with [spotless](https://github.com/diffplug/spotless/tree/main/plugin-gradle#ktlint)?

Not as far as I can tell, since experimental ktlint flags can't be enabled. (https://github.com/diffplug/spotless/issues/409)

**Comment by krisgerhard** [comment]
This feature seems to break `no-unused-imports` rule. I'm getting "unused import" error for all imports that are used on trailing comma lines exclusively

**Comment by hongjisung** [comment]
I have same `unused-imports` rule break error with `-F --experimental` option

I run `ktlint -F --experimental` three times then the project pass the test and fixed correctly
But, first and second trials are failed with only some modified files

**Comment by krisgerhard** [comment]
I created a failing test-case and new issue: https://github.com/pinterest/ktlint/issues/1367

**Comment by mikaello** [comment]
In [Ktlint v0.46.0](https://github.com/pinterest/ktlint/releases/tag/0.46.0), `trailing-comma` was promoted from the experimental ruleset to the standard ruleset. This should now work without the `--experimental` flag:

```editorconfig
[*.{kt,kts}]
ij_kotlin_allow_trailing_comma_on_call_site=true
ij_kotlin_allow_trailing_comma=true
```

If you use the [ktlint-maven-plugin](https://github.com/gantsign/ktlint-maven-plugin), you must use `v1.14.0` or later, example:

```shell
$ mvn com.github.gantsign.maven:ktlint-maven-plugin:1.14.0:format
```

**Comment by bgolob** [comment]
> In [Ktlint v0.46.0](https://github.com/pinterest/ktlint/releases/tag/0.46.0), `trailing-comma` was promoted from the experimental ruleset to the standard ruleset. Adding this to your `.editorconfig` should now work without having to specify `--experimental` flag when running Ktlint:
> 
> ```editorconfig
> [*.{kt,kts}]
> ij_kotlin_allow_trailing_comma_on_call_site=true
> ij_kotlin_allow_trailing_comma=true
> ```
> 
> If you use the [ktlint-maven-plugin](https://github.com/gantsign/ktlint-maven-plugin), you must use `v1.14.0` or later, example:
> 
> ```shell
> $ mvn com.github.gantsign.maven:ktlint-maven-plugin:1.14.0:format
> ```

I enabled both of the settings for trailing commas. I'm using  Ktlint Gradle plugin version 11.0.0. When I run ./gradlew ktlintcheck
I'm not getting any errors for missing trailing commas. 

For example here the trailing comma is missing and ktlint didn't throw any errors

`data class AdModel(
    val duration: Int,
    val type: String,
    val url: String
)`

**Comment by mikaello** [comment]
> I enabled both of the settings for trailing commas. I'm using Ktlint Gradle plugin version 11.0.0. When I run ./gradlew ktlintcheck
> I'm not getting any errors for missing trailing commas.

I am not familiar with Ktlint Gradle plugin, but it seems that the newest version of that plugin, [v11.0.0](https://github.com/JLLeitschuh/ktlint-gradle/releases/tag/v11.0.0), is [using Ktlint v0.43.2](https://github.com/JLLeitschuh/ktlint-gradle/blob/v11.0.0/plugin/gradle/libs.versions.toml#L3). Between Ktlint [[v0.43.0](https://github.com/pinterest/ktlint/releases/tag/0.43.0), v0.46.0), the trailing comma rule was only present in the experimental ruleset.

Skimming trough the issue list of Ktlint Gradle plugin it is at least issues for upgrading Ktlint, and problems with that, e.g.:
* https://github.com/JLLeitschuh/ktlint-gradle/issues/589
* https://github.com/JLLeitschuh/ktlint-gradle/issues/607

You should follow those issue to get notified when Ktlint Gradle plugin is compatible with newer versions of Ktlint.
